src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.IncrementalMemberEditAnalyzer.cs

```diff
@@ -126,47 +127,47 @@ public void UpdateDocumentWithCachedDiagnostics(Document document)
                 }
 
                 async Task ExecuteCompilerAnalyzerAsync(
-                    (DiagnosticAnalyzer analyzer, DocumentAnalysisData existingData, bool spanBased)? compilerAnalyzerData,
+                    (AnalyzerWithState analyzerWithState, bool spanBased)? compilerAnalyzerData,
                     ImmutableArray<TextSpan> oldMemberSpans,
                     PooledDictionary<DiagnosticAnalyzer, ImmutableArray<DiagnosticData>> builder)
                 {
                     if (!compilerAnalyzerData.HasValue)
                         return;
 
-                    var (analyzer, existingData, spanBased) = compilerAnalyzerData.Value;
+                    var (analyzerWithState, spanBased) = compilerAnalyzerData.Value;
                     var span = spanBased ? changedMember.FullSpan : (TextSpan?)null;
                     executor = executor.With(analysisScope.WithSpan(span));
-                    var analyzerAndExistingData = SpecializedCollections.SingletonEnumerable((analyzer, existingData));
-                    await ExecuteAnalyzersAsync(executor, analyzerAndExistingData, oldMemberSpans, builder).ConfigureAwait(false);
+                    var analyzersWithState = SpecializedCollections.SingletonEnumerable(analyzerWithState);
```

Comment:
>SpecializedCollections.SingletonEnumerable [](http://example.com/codeflow?start=45&length=42)

There are cool little gifts like this that I'm discovering each day in Roslyn. Out of curiosity, it doesn't look like this implements a struct based enumerator, do you know why?

Reply:
it's intended to be passed to something that takes an IEnumerable, so a struct-based-enumerator wouldn't help since it's going to be boxed anyways.

Reply:
I see. Couldn't ExecuteAnalyzersAsync be modified to take in an ArrayBuilder instead of IEnumerable to get the struct-based enumerator? (with the disadvantage being this codepath needs to get an arraybuilder from the pool instead of creating a singletonenumerable?)

Reply:
anything internal to us can def be updated.  it's only a challenge if it's an existing public entrypoint.

Reply:
It's defined inside this method, I guess the question I had was whether such a change would be desirable.

Reply:
yes.  if we control all the inputs, and all the methods are internal to us, going away from interface-based-collections is generally a good thing.  The primary deciding factor for me is if this is just like a single top level alloc that then is efficient underneath, or if this is a helper that will continually many times when doing work.  If the former, there's little motivation to change.  If the latter, it is def nicer to not have that constant stream of garbage.